Change relative import to extensioned variations for Node16 and NodeNext type support

datner opened the pull request:
- chore(msal-node): add $schema props for schema validation
- test(msal-node): add support for .js extension import
- build(msal-node): remove redundant pass
- feat(msal-node): change all relative imports to .js

I've made meaningful commits, so I recommend you review them instead of all the files at once.

The other solutions are:
1. to manually transform d.ts files after the build, which is a great way to introduce both complexity and brittleness to the project
2. use self-imports, allowing many entry-points `import { Configuration } from "@azure/msal-node/config/Configuration"` letting typescript and node to each resolve it in the way they understand. But this is a massive design-shift
3. using package.json `"imports"` entry. Just changing one problem for another, not a real solution and doesn't scale
4. emitting 2 `d.ts` versions, `d.mts` and `d.cts`. Thats possible but just.. redundant?

There are a few oddities, like the meme extensions importing only within their meme domain, but `d.ts` referencing non-existent `.js` files. Which is fine because it will just direct _Typescript_ to the correct `d.ts` and there isn't an interface difference between cjs and esm. Also `jest` had to be adjusted since it never got the memo that 3 LTS versions ago nodejs started supporting and advocating for `.js` extensions. If we ditch one of the meme extensions then this would be visually less scary, though it should be just as safe as it is now.

closes #6377

personal note:
Please don't bike-shed this pr for months, yes it would probably break node versions before 12 but come on. [even 16 is literally days away from deprecation](https://github.com/nodejs/release#release-schedule), it would be absolutely unreasonable.. This is over all a small change with a visually large foot print

datner committed:
chore(msal-node): add $schema props for schema validation

it's free and cool

datner committed:
test(msal-node): add support for .js extension import

absurd that this is not built in

datner commented:
@microsoft-github-policy-service agree

datner commented on lib/msal-node/jest.config.cjs:
```diff
@@ -19,5 +19,12 @@ module.exports = {
+     *  https://github.com/kulshekhar/ts-jest/issues/1057
+     */
+    moduleNameMapper: {
+      "^(\\.\\.?\\/.+)\\.js$": "$1"
```
this just morphs `"./some/relative/path.js"` to `"./some/relative/path"`

tnorling replied:
Could you please add this explanation to the comment above?

datner replied:
Ofc

datner commented on lib/msal-node/package.json:
```diff
@@ -1,4 +1,5 @@
 {
+  "$schema": "https://json.schemastore.org/package.json",
```
autocompletion and instant feedback is cool

datner commented on lib/msal-node/rollup.config.js:
```diff
@@ -5,17 +5,18 @@
 
 import { nodeResolve } from "@rollup/plugin-node-resolve";
 import typescript from "@rollup/plugin-typescript";
+import { defineConfig } from "rollup"
```
btw prettier does not cover these files

sameerag replied:
We can change that if it helps.

datner replied:
@sameerag  I just ran the prettier command manually, just raising a flag 😄

datner commented on lib/msal-node/rollup.config.js:
```diff
@@ -5,17 +5,18 @@
 const fileHeader = `${libraryHeader}\n${useStrictHeader}`;
 
-export default [
+export default defineConfig([
```
this does not change the config, only adds types

datner commented on lib/msal-node/rollup.config.js:
```diff
@@ -62,11 +44,9 @@ export default [
         ],
         plugins: [
             typescript( {
-                typescript: require("typescript"),
```
this is not required, its the default behavior

datner commented on lib/msal-node/rollup.config.js:
```diff
@@ -5,17 +5,18 @@
         // for cjs build
         input: "src/index.ts",
-        output: {
+        output: [{
```
we can can do multiple output instead of multiple bundling because everything else is identical

tnorling reviewed (commented):
Thanks for the PR! Could you please run `npm run beachball:change` from the root of the repo to generate changefiles

tnorling commented on lib/msal-node/src/index.ts:
```diff
@@ -118,4 +118,4 @@ export {
     AppTokenProviderResult,
 } from "@azure/msal-common";
```
Do these changes need to be made for msal-common exports as well?

datner replied:
`msal-common` is still `cjs`, once you change it to `"type": "module"` then the types will break and you'll have to move to the `.js` extension. At least thats as far as I understand

datner replied:
This does also mean that if you remove `"type": "module"` from the distributed package.json, this would also solve the issue and with the smallest footprint. But I would consider this a hack more than a solution

tnorling replied:
All of our libraries export both cjs and esm

datner replied:
But they share a `package.json`, take a peek at the one both msal-common and msal-nodes ship.
This is the difference between them, and what broke typescript

datner replied:
I'll be more specific. `msal-common` is still _written_ as a commonjs module for the simple fact that its type is not set to `"module"`-- So it defaults to `"commonjs"`. It and the rest are bundled for both `esm` and `cjs`, but the module type stays the same. Node decides which version to take wether you `import` or `require`, as outlined in your `"exports"` but can chew though both either way. the typescript type checker on the other hand, from what I understand, does not give 2 licks about how much effort we go though to bundle things universally, it knows either cjs _or_ esm. Depending on what the type of the module is-- which is decided by the package.json `"type"` field.

I'm sure you already know everything I said here, and I'm also sure I made mistakes while oversimplifying the issue, but thats the long version of why msal-common doesn't _need_ to do this yet, though it probably should

datner committed:
build(msal-node): simplify bundle process

also format the file with prettier
also also add type safety

datner committed:
feat(msal-node): change all relative imports to .js

This is much *much* safer and more reliable than posthumously
in some transforation hook. This also does not break current behavior so....

datner committed:
chore: changeset

datner commented:
No clue how coverage changed, there was literally 0 logical changes (tests or otherwise) 
This PR introduces a 20% improvement to build speed from roughly 1.3s to about 1s 😎 

I rebased from dev, formatted, and fixed conflicts

datner commented on lib/msal-node/rollup.config.js:
```diff
@@ -5,68 +5,49 @@
             dir: "dist",
             format: "cjs",
             entryFileNames: "[name].cjs",
             preserveModules: true,
             preserveModulesRoot: "src",
```
seeing as the only valid entry-point for the module is `@azure/msal-node`, I don't understand why these are used

tnorling commented:
Some of our CI tasks don't have permission to run on external PRs so I've pulled this into an internal branch and have opened #6399
